### ISSUE TITLE
Use string endpoints for `twistd mail` server address related options

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4739]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4739|
|Type|enhancement|
|Created|2010-11-15 00:47:15Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-mail-endpoints-4739-2|

`twistd mail` takes several options which define the address of a listening server:  `pop3`, `pop3s`, `smtp`.  These should accept values to be passed to `twisted.internet.endpoints.serverFromString` instead (this will make `pop3s` redundant, but it should be deprecated for a while according to the backwards compatibility policy).

Attachments:

* [ticket-4739.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/110/110758e6cb24ac07067a72a8d718dbafd5f5261c/dc986171735617df236face924bf657f2c85af17.patch) (10498 bytes) - added by tpratt on 2011-01-17 01:13:54Z - multiple smtp ports
* [ticket-4740.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/110/110758e6cb24ac07067a72a8d718dbafd5f5261c/bfa6b08da3f374c5da0a4565e4a28a829c938185.patch) (27 bytes) - added by tpratt on 2011-01-17 01:15:17Z - the next best thing to deleting this patch
* [4739.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/110/110758e6cb24ac07067a72a8d718dbafd5f5261c/e8f09c32232b4436a9d2c00d32d43e8204c5f5fe.patch) (10619 bytes) - added by tpratt on 2011-02-24 02:11:57Z - 
* [4739.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/110/110758e6cb24ac07067a72a8d718dbafd5f5261c/8c4a910f31dc04264d6b04499727fa33d2ceda76.patch) (12204 bytes) - added by tpratt on 2011-03-06 20:50:25Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4739 4739
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_twistd_mail_endpoints_4739_2 branches/twistd-mail-endpoints-4739-2
branch_author__exarkun__tpratt exarkun, tpratt
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1289782035000000 1289782035000000
changetime__1300724258000000 1300724258000000
version__None None
owner__exarkun exarkun
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|tpratt commented|
|-|-|

The != '0' may look strange at first, but I can't use usage.portCoerce since it might get a string that would raise a ValueError when it tries to create an int().

I'm a little unsure of the validity of "':' in description" as the test in toStrport().

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tpratt removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|tpratt commented|
|-|-|

There are a couple of questions with this patch.

1. --certificate causes SSL to be enabled on /all/ smtp ports.  That should be acceptable since --certificate is only retained for backward compatibility and the older versions only support one smtp listener.

2. postOptions checks that the file passed to --certificate exists.  Should I import the strport parser, get the privateKey parameter and validate that too?  It seems to defeat the generalness of strports to do that.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|tpratt commented|
|-|-|

oops, I attached the wrong patch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Replying to [tpratt](#note_5):
> oops, I attached the wrong patch.

Deleted your old patch..

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30878]) Branching to 'twistd-mail-endpoints-4739'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30879]) Apply 4739.patch

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to tpratt|
|-|-|

Thanks!  I like the direction this is going in.  Just a couple comments:

  1. Lots more tests needed.  Also it looks like you consumed `testPasswordfileDeprecation` to create `testPop3sOptionDeprecation`.  Presumably that was an accident? :)  The existing test should remain...
  1. With regards to `toStrport`, don't forget that you can just construct an endpoint yourself.  You don't need to format a string and then parse it.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|tpratt commented|
|-|-|


>   2. With regards to `toStrport`, don't forget that you can just construct an endpoint yourself.  You don't need to format a string and then parse it.

What I'm trying to accomplish with toStrport is to be able to provide backward compatibility with people who pass a bare port number as well as allow people to pass a strport going forward.  I don't really parse anything, but I think I need to decide if what they passed is a bare port or a strport.  Maybe I should just deprecate --pop3 and --smtp (they always accept just a bare port) and introduce new options for strports?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> What I'm trying to accomplish with toStrport is to be able to provide backward compatibility

Backwards compatibility is good.  I'm thinking more about the way string _formatting_ is used in it, though.  It would be nice to avoid things like:

```
    "tcp:%s" % description
```

and instead do things like:

```
    TCP4ServerEndpoint(reactor, int(description))
```

In other words, I think you know that `description` is a string giving a TCP port number in this case, so treat it as such and construct the `TCP4ServerEndpoint` directly, instead of creating a string that can be parsed to construct one.

_34. The string is a stark data structure and everywhere it is passed there is much duplication of process. It is a perfect vehicle for hiding information. _

> Maybe I should just deprecate --pop3 and --smtp 

I think the direction you've gone is good.  Maybe the port-number-only case of these options should be more visibly deprecated though (ie, emit a warning as well as documenting it to be deprecated).  Then we can eventually have the code be purely endpoint based instead of having to guess.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|tpratt removed owner|
|-|-|

added bunches of tests

realized my old patch was against an out-of-date working copy.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_16"></a>|@moshez set owner to @exarkun|
|-|-|

Comments:
PROTO_DISABLED should be _PROTO_DISABLED, because it really should be internal-only.
Why is line 128 have an extra newline?
Ditto line 53?

I think this can be checked in even with these things.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30978]) Branching to 'twistd-mail-endpoints-4739-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30979]) Apply 4739.2.patch 

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [30980]) Fix some coding-standard type issues, address review feedback

  * Make PROTO_DISABLED private
  * delete trailing whitespace
  * indent multi-line function calls consistently
  * two blank lines between methods, three between classes
  * fix copy/paste test docstrings, write docstrings that actually describe what is being tested
  * delete "Test that" prefix from test method docstrings
  * switch to `assert*` methods instead of `failUnless*` methods
  * Rename test method names to use `test_` prefix

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

I fixed some boring stuff, described in the previous comment.  There are a couple more involved changes I'd like to see as well though:

  1. The test suite triggers some deprecations that don't get flushed.  It needs to avoid this so they don't show up in trial's output.

Oops, I forget what the rest is, and now it's pretty late.  Submitting this though it is incomplete, so at least that one point isn't lost.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

I did a bit more.  I replaced `--pop3 0` with `--no-pop3`.  There are some deprecations being emitted by the test suite that need to be fixed, and there might be some missing test coverage, I'm not sure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun removed owner|
|-|-|

And a little more, suppressing those deprecations and adding a bit more test coverage.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twistd-mail-endpoints-4739-2)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph set owner to @exarkun|
|-|-|

Coverage for `twisted.mail.tap` still looks pretty sad, but not for the stuff added in this branch.  Please make a ticket for improving that.

Also, the deprecated 'certificate' option should say 'see deprecated pop3s option'.  Right now, it reads as if you should use `privateKey` if you were using `--pop3s`, but just `--pop3` if you were using `--certificate`.  Obviously that's nonsensical, so we should be explicit about it.

Otherwise the contents of the branch look pretty good.  Please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [31288]) Try to clarify the deprecation situation with --certificate

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `closed`|
|-|-|

(In [31290]) Merge twistd-mail-endpoints-4739-2

Author: tpratt, exarkun
Reviewer: exarkun, glyph
Fixes: [#4739](https://github.com/twisted/twisted/issues/4739)

Accept endpoint descriptions in `twistd mail` for the POP3 and SMTP listener address
configuration.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `reopened`|
|-|-|

(In [31358]) Revert r31290 - test failure on Windows, which lacks "/dev/null"

Reopens [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [31359]) Use os.devnull instead of "/dev/unull"

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [31360]) Forget about devnull, just use a file.

refs [#4739](https://github.com/twisted/twisted/issues/4739)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun set status to `closed`|
|-|-|

(In [31361]) Merge twistd-mail-endpoints-4739-2

Author: tpratt, exarkun
Reviewer: exarkun, glyph
Fixes: [#4739](https://github.com/twisted/twisted/issues/4739)

Accept endpoint descriptions in `twistd mail` for the POP3 and SMTP listener address
configuration.

Re-merge fixes a problem in the new tests, which previously used /dev/null which
is problematic on Windows.
